pkg/virt-handler/rest/console.go

```diff
@@ -54,8 +61,60 @@ func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiI
 		vncStopChans:         make(map[types.UID](chan struct{})),
 		serialLock:           &sync.Mutex{},
 		vncLock:              &sync.Mutex{},
+		usbredirLock:         &sync.Mutex{},
 		vmiInformer:          vmiInformer,
+		usbredir:             make(map[types.UID]UsbredirHandlerVMI),
+	}
+}
+
+func (t *ConsoleHandler) USBRedirHandler(request *restful.Request, response *restful.Response) {
+	vmi, code, err := getVMI(request, t.vmiInformer)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed to retrieve VMI")
+		response.WriteError(code, err)
+		return
+	}
+
+	// For simplicity, we handle one usbredir request at the time, for all VMIs
+	// handled by virt-handler
+	t.usbredirLock.Lock()
+
```

Comment:
I suggest we utilize go's `defer` and use `defer t.usbredirLock.Unlock()` here.

Reply:
Right, I tried that initially but IIRC, because t.stream() doesn't return while the stream is active, the Unlock() didn't get called. The main goal of this lock was to avoid a race request to the same `virt-usbredir-$label` socket.

Reply:
Got you.
This is a very interesting case where `defer` is indeed problematic. I can think of 3 plausible alternatives:

1. Keep what you did.
2. Use `defer t.usbredirLock.Unlock()`. In the end of the function, do this:
    ```
	t.usbredirLock.Unlock()
	t.stream(vmi, request, response, unixSocketPath, stopChan, cleanup)
	t.usbredirLock.Lock()
    ```
    This is the safest way to go since if we return from one of the `if` branches we can be sure that the lock is released. On the other hand I would agree that this has ugliness to it and it's not super obvious. It also might hurt performance since you need to aquire and release the lock for no actual reason. But since, IMHO, writing safe code is critical I would consider this option.
3. Run `t.stream()` in a background thread, i.e. `go t.stream(...)`. I'm just not sure that the stream function can run safely in parallel. I have a feeling this is more complicated that it sounds.

Up to you :)

Reply:
I think I tried (3) too, I'll double check just to be sure. I don't mind going with option (2) if you think it is safer. I'll try it out and submit the changes later.

Reply:
(3) does not work as returning that function kills the stream. This is hard to workaround I think, because it is a webservice callback from cmd/virt-handler/virt-handler.go